crates/matrix-sdk/tests/integration/event_cache.rs

```diff
@@ -0,0 +1,164 @@
+use std::time::Duration;
+
+use assert_matches2::{assert_let, assert_matches};
+use matrix_sdk::{
+    event_cache::{EventCacheError, RoomEventCacheUpdate},
+    test_utils::{logged_in_client, logged_in_client_with_server},
+};
+use matrix_sdk_common::deserialized_responses::SyncTimelineEvent;
+use matrix_sdk_test::{
+    async_test, sync_timeline_event, EventBuilder, JoinedRoomBuilder, SyncResponseBuilder,
+};
+use ruma::{
+    events::{
+        room::message::{MessageType, RoomMessageEventContent},
+        AnySyncMessageLikeEvent, AnySyncTimelineEvent,
+    },
+    room_id, user_id,
+};
+use tokio::time::timeout;
+
+use crate::mock_sync;
+
+fn assert_event_matches_msg(event: &SyncTimelineEvent, expected: &str) {
+    let event = event.event.deserialize().unwrap();
+    assert_let!(
+        AnySyncTimelineEvent::MessageLike(AnySyncMessageLikeEvent::RoomMessage(message)) = event
+    );
+    let message = message.as_original().unwrap();
+    assert_let!(MessageType::Text(text) = &message.content.msgtype);
+    assert_eq!(text.body, expected);
+}
+
+#[async_test]
+async fn test_must_explicitly_subscribe() {
+    let client = logged_in_client(None).await;
+
+    let event_cache = client.event_cache();
+
+    // If I create a room event subscriber for a room before subscribing the event
+    // cache,
+    let room_id = room_id!("!omelette:fromage.fr");
+    let result = event_cache.for_room(room_id).await;
+
+    // Then it fails, because one must explicitly call `.subscribe()` on the event
+    // cache.
+    assert_matches!(result, Err(EventCacheError::NotSubscribedYet));
+}
+
+#[async_test]
+async fn test_cant_subscribe_to_unknown_room() {
+    let client = logged_in_client(None).await;
+
+    let event_cache = client.event_cache();
+
+    // First, subscribe to the event cache.
+    event_cache.subscribe().unwrap();
+
+    // If I create a room event subscriber for a room unknown to the client,
+    let room_id = room_id!("!omelette:fromage.fr");
+    let result = event_cache.for_room(room_id).await;
```

Comment:
Would it make sense to move this method under the `Room` struct? This way the error can't happen.

Reply:
Sorry, which method is `this method`? This tests checks that calling `for_room` fails for unknown rooms, so we're not calling a room-specific method here yet.

Reply:
Sorry, thought my ramblings would have been clearer. What I mean is, could we move the method `EventCache::for_room(&self, room_id: &RoomId)` to be under the `Room` object.

So you would have to call `Room::get_event_cache(&self)` instead of `EventCache::for_room()`. This way people need to fetch the room and the error case you are checking for in this test ceases to exist.

Reply:
That's a good idea! Will do in a follow-up.